controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -1853,8 +1854,14 @@ public TaskBase copyWithContext(Context context) {
 
     @Override
     public void close() throws Exception {
-        if (eventHelper != null) {
-            eventHelper.close();
+        if (eventHelperFuture.isDone()) {
+            eventHelperFuture.thenAccept(eventHelper -> {
+                if (eventHelper != null) {
+                    eventHelper.close();
+                }
+            }).join();
+        } else {
+            eventHelperFuture.cancel(true);
```

Comment:
This isn't going to have any effect. 

Furthermore, above (in `initialize`) when you set the value, that call will have no effect if this future is already completed (whether successfully or with exception). That means that the `EventHelper` object you create will be a source of resource leaks as you never close it.

Please rewrite your code to handle that situation.

Reply:
we should do the following:
create eventHelper and set it in a lock protected variable and use a boolean to indicate if it needs to be set or not.
init method:
```
synchronize(lock) {
if(toSetEventHelper) {
eventhelper = new EventHelper();
toSetEventHelper = false;
}
}
// after we release the lock, if close was invoked then eventhelper may have been closed. but completing the future with that is safe (and its ok even if eventhelperfuture was cancelled in close method).
if (eventHelper != null) {
eventHelperFuture.complete(eventHelper);
}

```
close method:
```
synchronize(lock) {
toSetEventHelper = false;
if (eventhelper != null) {
eventHelper.close();
}
}
eventhelperfuture.cancel(true);
```

Reply:
Changes are done as per suggested.